faunadb-java-dsl/src/main/java/com/faunadb/client/types/Decoder.java

```diff
@@ -0,0 +1,233 @@
+import com.faunadb.client.types.Value.NullV;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.Type;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.lang.String.format;
+import static java.lang.reflect.Modifier.isAbstract;
+
+/**
+ * FaunaDB {@link Value} to object decoder.
+ */
+public final class Decoder {
+  private Decoder() {
+  }
+
+  private final static ConcurrentHashMap<Class<?>, Codec<?>> CODECS = new ConcurrentHashMap<>();
+
+  static {
+    Codecs.initDefault(CODECS);
+  }
+
+  /**
+   * Decode a FaunaDB {@link Value} to a specified type.
+   * <p>
+   * <p>This method is useful if you need to decode
+   * collections types.</p>
+   * <p>
+   * <pre>{@code
+   *     ArrayV array = new ArrayV(Arrays.asList(new StringV("str1"), new StringV("str2")));
+   *
+   *     Result<List<String>> listStrings = Decoder.decode(array, Types.arrayListOf(String.class));
+   *     Result<Set<String>> setStrings = Decoder.decode(array, Types.hashSetOf(String.class));
+   * }</pre>
+   * <p>
+   * See {@link Types} for more information.
+   *
+   * @param <T>     The return type of the method.
+   * @param value   The FaunaDB {@link Value} to be decoded.
+   * @param dstType The {@link Type} in which value should be decoded.
+   * @return A {@link Result} instance of type {@link T}
+   */
+  public static <T> Result<T> decode(Value value, Type dstType) {
+    return (Result<T>) decodeImpl(value, Types.of(dstType));
+  }
+
+  /**
+   * Decode a FaunaDB {@link Value} to a specified type.
+   * <p>
+   * <p>Use this method to decode user defined types like:</p>
+   * <p>
+   * <pre>{@code
+   *     Result<User> user = Decoder.decode(new ObjectV(...), User.class);
+   * }</pre>
+   * <p>
+   * <p>It's possible to decode primitive types and array of primitive types:</p>
+   * <p>
+   * <pre>{@code
+   *     Result<String> string = Decoder.decode(new StringV("string"), String.class);
+   *     Result<long> longValue = Decoder.decode(new LongV(10), long.class);
+   *     Result<int> intValue =  Decoder.decode(new LongV(10), int.class);
+   *
+   *     Result<int[]> arrayInt = Decoder.decode(new ArrayV(...), int[].class);
+   * }</pre>
+   *
+   * @param <T>     The return type of the method.
+   * @param value   The FaunaDB {@link Value} to be decoded.
+   * @param dstType The {@link Class} in which value should be decoded.
+   * @return A {@link Result} instance of type {@link T}
+   */
+  public static <T> Result<T> decode(Value value, Class<T> dstType) {
+    return (Result<T>) decodeImpl(value, Types.of(dstType));
+  }
+
+  static Result<?> decodeImpl(Value value, SimpleType dstType) {
+    Class<?> rawType = dstType.getRawClass();
+
+    if (rawType == Object.class)
+      return Result.success(value == NullV.NULL ? null : value);
```

Comment:
Interesting. Our decoders don't decode `null` so that we keep the contract with the `Result` interface. 

For example:
```java
// { name: null }
value
  .at("name")
  .to(Codec.STRING)
  .map(name -> name.toUpperCase())
  .getOrElse("No name");
```

If we decode nulls, the `map` transformation would result in a NPE.

Although, it seems like it makes sense to decode to null when decoding for a specific type. If that same object was decoded to a class where `name` is a property of type `String`, it does make sense to decode to `null`. Therefore, I'm starting to feel like it also make sense to decode to null if the desired type is an `Object`. I think that's what most developers would expect.

It seems like we need to respect the default values in case of nulls. Something like `decodeImp(..).getOrElse(defaultForType(..))`.